common/models/transducer/multi_stager.py

```diff
@@ -0,0 +1,284 @@
+  stage_epochs=15,
+  fixed_path=True,
+  alignment_topology=rnnt_topology)
+
+# Multi stage training with pretraining
+get_network = MultiStager([st1, st2]).get_network
+
+  TODO:
+    - [ ] How to save the information about the differences between alignments of different topologies.
+    - [ ] Make sure that the alignments correspond to the dataset used(sequence_ordering, ..)
+    - [ ] Reset option
+    - [ ] How to define loops? Instead of creating Stages manually (could do a for loop)
+"""
+
+from __future__ import annotations
+from ast import Str
+from typing import Dict, Any, List, AnyStr
+
+from returnn.tf.util.data import Data, DimensionTag
+from returnn.config import get_global_config
+from ...training.switchout import switchout_target
+from ...datasets.interface import TargetConfig
+from .topology import Topology, rna_topology, rnnt_topology
+
+import tensorflow as tf
+import sys
+import os
+
+
+
+class Context:
+  def __init__(self, task: str, target: TargetConfig, model: str,
+               name: str, alignment_topology: Topology = rnnt_topology):
+    self.task = task
+    self.train = (task == "train")
+    self.search = (task != "train")
+    self.target = target
+    self.align_dir = os.path.dirname(model)
+    self.name = name
+    self.num_labels_nb = target.get_num_classes()
+    self.num_labels_wb = self.num_labels_nb + 1
+    self.blank_idx = self.num_labels_nb
+    self.alignment_topology = alignment_topology
+
+
+def make_align(net: Dict[str, Any],
+               epoch: int,  # maybe required
+               extend: bool,
+               ctx: Context,
+               output: str = "output",
+               output_log_prob: str = "output_log_prob_wb",
+               encoder: str = "encoder",
+               target: TargetConfig = None):
+  """
+  Here we assume that the decoder is a recurent network(with unit) called `output`.
+  In the "unit" `output_log_prob` should define the the log distribution over the whole vocab inkl blank.
+  Otherwise "base:{encoder}" which represent the output of the encoder should be provided.
+  This function extends the "unit" of the decoder with logic to create and dump fullsum alginment in .hdf files.
+  Requires:
+    output/unit
+      output_log_prob: log distribution over the whole vocab inkl blank
+      f"base:data:{target}": targets of the sequence
+      base:encoder: output of the encoder
+    rna_fullsum_alignment: function that performs the alignment and returns for e.g [BxT] for rna alignm.
+    extend: if True the already existing .hdf alignments are extended
+    ctx: holds information such as the label topology, target and path to be used for .hdf files
+
+  Durign this step 1 subepoch passes.
+  """
+  align_dir = ctx.align_dir
+  name = ctx.name
+  if not target:
+    target = TargetConfig.global_from_config()
+  subnet = net[output]["unit"]
+  subnet[output_log_prob]["is_output_layer"] = True
+  subnet["fullsum_alignment"] = {
```

Comment:
Why do you call it "fullsum_alignment"? This is not doing the full sum, but instead it does the max, or argmax.

Reply:
is max_alignment ok?

Reply:
Yes.